[#14] 회원가입, 로그인, 마이페이지 조회, 비밀번호 수정, 의사 이미지 수정, 토큰 재발급 기능 생성

suyeonkim56 opened the pull request:
#14 
회원가입, 로그인, 마이페이지 조회, 비밀번호 수정, 의사 이미지 수정, 토큰 재발급 기능 생성

suyeonkim56 committed:
Feat : 회원가입, 로그인 기능 추가( #14 )

suyeonkim56 committed:
Feat : 마이페이지 조회 기능 추가 ( #14 )

suyeonkim56 committed:
Feat : 비밀번호 수정, 의사 이미지 수정 기능 추가 ( #14 )

suyeonkim56 committed:
Feat : 토큰 재발급 기능 추가 ( #14 )

suyeonkim56 committed:
Merge branch 'dev' into feat/login

suyeonkim56 committed:
Refactor : 생성자 수정 ( #14 )

suyeonkim56 committed:
Merge branch 'dev' into feat/login

suyeonkim56 committed:
Merge branch 'dev' into feat/login

tggKim commented on src/main/java/com/example/docconneting/common/exception/constant/ErrorCode.java:
이 부분 상태코드에 대해서 전체적으로 수정이 조금 필요할 것 같다는 생각이 듭니다

tggKim commented on src/main/java/com/example/docconneting/domain/Auth/dto/request/UserRefreshTokenRequestDto.java:
@NotNull 보다는 @NotBlank 가 어떨지 여쭤보고싶습니다

tggKim commented on src/main/java/com/example/docconneting/domain/Auth/dto/request/UserSignUpRequestDto.java:
여기서도 userRole에 @NotBlank 가 어떨지 여쭤보고싶습니다

tggKim commented on src/main/java/com/example/docconneting/domain/Auth/dto/request/UserSigninRequestDto.java:
여기서도 @NotBlank가 어떨지 여쭤보고 싶습니다

tggKim commented on src/main/java/com/example/docconneting/common/config/JwtUtil.java:
```diff
@@ -0,0 +1,78 @@
+    public String createRefreshToken(Long userId) {
+        Date date = new Date();
+
+        return BEARER_PREFIX +
```
여기서 Bearer를 안붙여도 될 것 같다는 생각이 듭니다

tggKim commented on src/main/java/com/example/docconneting/domain/Auth/service/AuthService.java:
```diff
@@ -0,0 +1,131 @@
+
+    //토큰 재발급
+    @Transactional
+    public UserRefreshTokenResponseDto refreshToken(AuthUser authuser, UserRefreshTokenRequestDto dto) {
```
리프레시 토큰이 jwt라서 만료기간이 있으니까
받아오는 리프레시 토큰이 만료되었는지의 여부를 확인하는 로직이 있으면 좋을 것 같습니다.

그리고 여기서 리프레시 토큰을 새롭게 생성을 후
-  레디스의 레프레시 토큰도 업데이트
- 새롭게 생성된 리프레시 토큰을 리턴하도록 하는게 좋을 것 같다고 생각이 듭니다

tggKim commented on src/main/java/com/example/docconneting/domain/user/entity/User.java:
```diff
@@ -11,13 +11,14 @@
 @Table(name = "users",
-        indexes = { @Index(name = "idx1", columnList = "isDeleted"),
-                    @Index(name = "idx2", columnList = "major, isDeleted") }
+        indexes = {@Index(name = "idx1", columnList = "isDeleted"),
```
인덱스 이름 정의하는 것에 대해 토의가 필요해 보입니다

tggKim commented on src/main/resources/application.yml:
여기서 시크릿 키에 대한 부분도 application-local.yml에 넣으시면 좋을 것 같습니다

kyn1013 commented on src/main/java/com/example/docconneting/common/filter/JwtFilter.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    // URI + HTTP Method 기반 화이트리스트
+    private boolean isWhiteList(String requestURI, String method) {
+        if (requestURI.startsWith("/api/v1/signup") || requestURI.startsWith("/api/v1/signin")) {
+            return true;
+        }
+
+        if ("GET".equals(method)) {
+            return requestURI.matches("^/api/v1/posts$") ||
+                    requestURI.matches("^/api/v1/posts/\\d+$") ||
+                    requestURI.matches("^/api/v1/posts/\\d+/comments$");
+        }
+
+        return false;
+    }
```
url 검증을 filter config가 아니라 화이트 리스트로 따로 메서드로 만드신 이유가 궁금합니다!

kyn1013 commented on src/main/java/com/example/docconneting/common/filter/JwtFilter.java:
```diff
@@ -0,0 +1,105 @@
+
+            chain.doFilter(request, response);
+
+        } catch (SecurityException | MalformedJwtException e) {
+            log.error("유효하지 않은 JWT 서명입니다.", e);
+            httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED, "유효하지 않은 JWT 서명입니다.");
+        } catch (ExpiredJwtException e) {
+            log.error("만료된 JWT 토큰입니다.", e);
+            httpResponse.sendError(HttpServletResponse.SC_UNAUTHORIZED, "만료된 JWT 토큰입니다.");
+        } catch (UnsupportedJwtException e) {
+            log.error("지원되지 않는 JWT 토큰입니다.", e);
+            httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "지원되지 않는 JWT 토큰입니다.");
+        } catch (Exception e) {
+            log.error("유효하지 않은 JWT 토큰입니다.", e);
+            httpResponse.sendError(HttpServletResponse.SC_BAD_REQUEST, "유효하지 않은 JWT 토큰입니다.");
```
저희가 만든 에러객체랑 응답코드로 수정이 필요할 것 같습니다!

kyn1013 commented on src/main/java/com/example/docconneting/common/resolver/AuthUserArgumentResolver.java:
```diff
@@ -0,0 +1,44 @@
+
+        // @Auth 어노테이션과 AuthUser 타입이 함께 사용되지 않은 경우 예외 발생
+        if (hasAuthAnnotation != isAuthUserType) {
+            throw new IllegalStateException("@Auth와 AuthUser 타입은 함께 사용되어야 합니다.");
```
저희가 만든 exception과 에러 코드 적용이 필요할 것 같습니다!

hyeonseokhoon commented on src/main/java/com/example/docconneting/common/exception/constant/ErrorCode.java:
```diff
@@ -11,6 +11,23 @@ public enum ErrorCode {
     // 403
     EMAIL_ALREADY_EXISTS(HttpStatus.BAD_REQUEST, "이미 존재하는 이메일입니다."),
+    PASSWORD_SAME_AS_OLD(HttpStatus.BAD_REQUEST,"기존 비밀번호와 동일한 비밀번호로 수정할 수 없습니다."),
+
```
private final String code; 추가하여

EMAIL_ALREADY_EXISTS("U001", HttpStatus.BAD_REQUEST, "이미 존재하는 이메일입니다."),
PASSWORD_SAME_AS_OLD("U002", HttpStatus.BAD_REQUEST,"기존 비밀번호와 동일한 비밀번호로 수정할 수 없습니다."),

이렇게 하면
BAD_REQUEST 상태가 어떤 오류에 의한 것인지 확인하기 좀더 쉬울것 같습니다.

hyeonseokhoon commented on src/main/java/com/example/docconneting/common/filter/JwtFilter.java:
```diff
@@ -0,0 +1,105 @@
+        if ("GET".equals(method)) {
+            return requestURI.matches("^/api/v1/posts$") ||
+                    requestURI.matches("^/api/v1/posts/\\d+$") ||
+                    requestURI.matches("^/api/v1/posts/\\d+/comments$");
```
access_token이 만료되어 refresh_token으로 token을 갱신하는 경우도 있으니,  whiteList나 어떤 방식으로든 만료된 토큰 또는 Authorization 이 빈경우 /api/v1/refresh 에 대한 접근이 가능하도록 해야 합니다.

kyn1013 commented on src/main/java/com/example/docconneting/domain/Auth/service/AuthService.java:
```diff
@@ -0,0 +1,131 @@
+                }
+
+                Major major = Major.of(dto.getMajor().toUpperCase());
+                yield User.of(
```
yield가 뭔지 설명 부탁드립니다!

hyeonseokhoon commented on src/main/java/com/example/docconneting/common/filter/JwtFilter.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public void destroy() {
```
코드가 super.destroy() 이니 딱히 override하지 않으면 super의 destroy를 사용 합니다.

wonchaebin commented on src/main/java/com/example/docconneting/domain/Auth/dto/request/UserSigninRequestDto.java:
sign up에서는 비번이랑 이메일을 not blank로 처리하셨는데 여기는 not null로 처리하신 이유가 궁금합니다!

kyn1013 commented on src/main/java/com/example/docconneting/domain/Auth/service/AuthService.java:
```diff
@@ -0,0 +1,131 @@
+
+        String newAccessToken = jwtUtil.createToken(user.getId(), user.getUserRole());
+
+        return UserRefreshTokenResponseDto.of(newAccessToken, refreshToken);
```
토큰을 재발급하면 레디스에 기존 리프레시 토큰은 삭제되고 새로운 리프레시 토큰은 레디스에 저장되는 건가요? 혹시 맞다면 그 부분도 코드 궁금합니다!!

wonchaebin commented on src/main/java/com/example/docconneting/domain/Auth/dto/response/UserRefreshTokenResponseDto.java:
```diff
@@ -0,0 +1,20 @@
+package com.example.docconneting.domain.Auth.dto.response;
+
+
+import lombok.Builder;
```
import 빌더 지워야할거 같습니다!

wonchaebin commented on src/main/java/com/example/docconneting/domain/Auth/dto/response/UserSignInResponseDto.java:
```diff
@@ -0,0 +1,19 @@
+package com.example.docconneting.domain.Auth.dto.response;
+
+import lombok.Builder;
```
여기도 지워주시면 될거 같습니다 import 빌더

wonchaebin commented on src/main/java/com/example/docconneting/domain/user/dto/response/UserMyPageResponseDto.java:
```diff
@@ -0,0 +1,19 @@
+package com.example.docconneting.domain.user.dto.response;
+
+import lombok.Builder;
```
빌더가 안지워진거 같습니다!

wonchaebin reviewed (approved):
코드 짜시느라 고생하셨습니다!

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/Auth/service/AuthService.java:
```diff
@@ -0,0 +1,131 @@
+
+        User user = switch (role) {
+            case DOCTOR -> {
+                if (dto.getMajor() == null) {
```
https://www.baeldung.com/spring-mvc-custom-validator
https://k3068.tistory.com/101

custom validation annotation 을 만들어 보면 어떨까요? 사용자 롤이 DOCTOR 인 경우 (@NotNull4Doctor) null 일 수 없다를 정의해 보는 거죠.

hyeonseokhoon commented on src/main/java/com/example/docconneting/domain/Auth/service/AuthService.java:
```diff
@@ -0,0 +1,131 @@
+
+        Map<String, String> message = new HashMap<>();
+        message.put("message", "회원 가입이 성공적으로 됐습니다");
+        return Response.of(message);
```
Response 가 어떤 포멧인지에 따라 다르겠지만 Map을 Response에 of 로 처리하는 것이 뭔가 어색해 보입니다.

suyeonkim56 committed:
Refactor : 코드 리뷰 기반 수정 ( #14 )

- 리프레시 토큰 발급할때 bearer 제거
- 권한 이름 에러 - 400으로 수정
- 비밀번호 입력 에러 - 400으로 수정
- 이미지 입력 - 400
- 근무 시작/종료시간 - 400
- string쪽은 @notblank로 valid하기
 - 어세스 토큰 재발급시 리프레시 토큰 만료 확인 및 리프레시 토큰도 재발급하고 레디스에 업데이트

- 시크릿키 local로 옮기고 옮긴거 슬랙에 보고

-@Auth 익셉션 서버에러익셉션으로 처리

suyeonkim56 committed:
Merge branch 'dev' into feat/login
